File: site/02-Linked-Lists/js/singly_linked_list.js
class Node {
    constructor(data) {
        this.data = data;
        this.next = null;
    }
}

class SinglyLinkedList {
    constructor() {
        this.head = null;
    }

    insertAtStart(data) {
        const newNode = new Node(data);
        newNode.next = this.head;
        this.head = newNode;
    }

    insertAtEnd(data) {
        const newNode = new Node(data);
        if (!this.head) {
            this.head = newNode;
            return;
        }
        let temp = this.head;
        while (temp.next) {
            temp = temp.next;
        }
        temp.next = newNode;
    }

    deleteNode(data) {
        if (!this.head) return;
        if (this.head.data === data) {
            this.head = this.head.next;
            return;
        }
        let temp = this.head;
        while (temp.next && temp.next.data !== data) {
            temp = temp.next;
        }
        if (temp.next) {
            temp.next = temp.next.next;
        }
    }

    display() {
        if (!this.head) {
            console.log("List is empty");
            return;
        }
        let temp = this.head;
        const result = [];
        while (temp) {
            result.push(temp.data);
            temp = temp.next;
        }
        console.log(result.join(" -> ") + " -> null");
    }
}

// Driver testing
const list = new SinglyLinkedList();
console.log("Inserting 10, 20, 30 at end:");
list.insertAtEnd(10);
list.insertAtEnd(20);
list.insertAtEnd(30);
list.display();

console.log("Inserting 5 at start:");
list.insertAtStart(5);
list.display();

console.log("Deleting 20:");
list.deleteNode(20);
list.display();

File: 02-Linked-Lists/java/SinglyLinkedList.java
public class SinglyLinkedList {
    private static class Node {
        int data;
        Node next;
        
        Node(int val) {
            this.data = val;
            this.next = null;
        }
    }

    private Node head;

    public SinglyLinkedList() {
        this.head = null;
    }

    public void insertAtStart(int val) {
        Node newNode = new Node(val);
        newNode.next = head;
        head = newNode;
    }

    public void insertAtEnd(int val) {
        Node newNode = new Node(val);
        if (head == null) {
            head = newNode;
            return;
        }
        Node temp = head;
        while (temp.next != null) {
            temp = temp.next;
        }
        temp.next = newNode;
    }

    public void deleteNode(int val) {
        if (head == null) return;
        if (head.data == val) {
            head = head.next;
            return;
        }
        Node temp = head;
        while (temp.next != null && temp.next.data != val) {
            temp = temp.next;
        }
        if (temp.next != null) {
            temp.next = temp.next.next;
        }
    }

    public void display() {
        if (head == null) {
            System.out.println("List is empty");
            return;
        }
        Node temp = head;
        while (temp != null) {
            System.out.print(temp.data + " -> ");
            temp = temp.next;
        }
        System.out.println("null");
    }

    public static void main(String[] args) {
        SinglyLinkedList list = new SinglyLinkedList();
        System.out.println("Inserting 10, 20, 30 at end:");
        list.insertAtEnd(10);
        list.insertAtEnd(20);
        list.insertAtEnd(30);
        list.display();
        
        System.out.println("Inserting 5 at start:");
        list.insertAtStart(5);
        list.display();
        
        System.out.println("Deleting 20:");
        list.deleteNode(20);
        list.display();
    }
}
